pkg/metrics/metrics.go

```diff
@@ -0,0 +1,222 @@
+// Copyright 2024 The Authors (see AUTHORS file)
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package metrics includes code specific to sending usage metrics.
+package metrics
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"strings"
+	"time"
+
+	"github.com/sethvargo/go-envconfig"
+
+	"github.com/abcxyz/abc-updater/pkg/optout"
+	"github.com/abcxyz/pkg/logging"
+)
+
+const (
+	installIDFileName     = "id.json"
+	maxErrorResponseBytes = 2048
+)
+
+type metricsConfig struct {
+	ServerURL string `env:"ABC_METRICS_URL, default=https://abc-updater-metrics.tycho.joonix.net"`
+}
+
+type options struct {
+	httpClient *http.Client
+	lookuper   envconfig.Lookuper
+	// Optional override for install id file location. Mostly intended for testing.
+	// If empty uses default location.
+	installIDFileOverride string
+}
+
+// Option is the metrics Client option type.
+type Option func(*options) *options
+
+// WithHTTPClient instructs the Client to use given http.Client when making calls.
+func WithHTTPClient(client *http.Client) Option {
+	return func(o *options) *options {
+		o.httpClient = client
+		return o
+	}
+}
+
+// WithLookuper instructs the Client to use given envconfig.Lookuper when
+// loading configuration.
+func WithLookuper(lookuper envconfig.Lookuper) Option {
+	return func(o *options) *options {
+		o.lookuper = lookuper
+		return o
+	}
+}
+
+// For testing. Can expose externally if a need arises in the future.
+func withInstallIDFileOverride(path string) Option { //nolint:unused
+	return func(o *options) *options {
+		o.installIDFileOverride = path
+		return o
+	}
+}
+
+// TODO: should Client be an interface so we can have a noop client returned for caller to use in case of error?
+// TODO: Alternatively, I could never return an error, and instead just set optOut if I encounter error in setup,
+// TODO: which is effectively a noop client.
```

Comment:
Thoughts, anyone?

Reply:
My two cents: aim for the API principle of making the right thing be the obvious thing and vice versa.

So `New()` could return an optout client unless the caller provides the "WithReturnErrors()" option.

And maybe rename the `optOut` field to `noop` to reflect that it's used for other things besides the opt-out case. And maybe make it a bool instead of *OptOutSettings.

Reply:
The interface should be `SendSync`, not `Client`, but seems reasonable. This can be a follow-up though.

Reply:
Hmm, I don't like the user having to deal with `error` in the signature for the case of not returning errors, though I do like the idea of them being opt-in

Reply:
What about this:

Always return a client, but if there is an error provided, make it an opt-out client, and app implementer could choose to handle error or not, but would never have a case where they run into a nill pointer issue.

Reply:
Seems fine to me.